### PR TITLE
feat: Add app.getLocaleCountryCode() method for region detection

### PR DESCRIPTION
**Pull request by zarubond**
##### Description of Change
On desktop operating systems the users can set different language and locale region. This allows the system to have en-CZ locale even through there is no czech region associated with english (the example results in English language with Czech data/number format/currency). However Chrome only presents region in locale based on system language.

This patch adds the API returning the OS region which can be then used for globalization. With the knowledge of region the applications would be able to show proper localization as any other native app.

##### Checklist
- [x] PR description included and stakeholders cc'd
- [x] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)

##### Release Notes
Notes: feat: provide user system's region with `app.getLocaleCountryCode()`

### CODE REVIEW

**Comment by alexeykuzmin** [comment]
@zarubond can you please rebase your branch onto the latest `electron:master`? A few CI fixes landed there in the last several days.

File: atom/browser/api/atom_api_app.cc [hunk 1]
```diff
@@ -874,6 +875,50 @@ std::string App::GetLocale() {
...
+    std::string::size_type rpos = locale.rfind('.');
+    if (rpos != std::string::npos)
+      locale = locale.substr(0, rpos);
+
+    rpos = locale.rfind('_');
+    if (rpos != std::string::npos && rpos + 1 < locale.size())
+      region = locale.substr(rpos + 1, 2);
+  }
+#endif
+  if (region.size() == 2)
```
Comment (codebytere) [inline atom/browser/api/atom_api_app.cc]: nit: this is probably cleaner as a ternary: 
return region.size() == 2 ? region : std::string();

File: atom/browser/api/atom_api_app.cc [hunk 2]
```diff
@@ -874,6 +875,50 @@ std::string App::GetLocale() {
...
+                      (LPWSTR)&locale_name,
+                      sizeof(locale_name) / sizeof(WCHAR)))
+    base::WideToUTF8(locale_name, wcslen(locale_name), &region);
+#elif defined(OS_MACOSX)
+  CFLocaleRef locale = CFLocaleCopyCurrent();
+  CFStringRef value = CFStringRef(
+      static_cast<CFTypeRef>(CFLocaleGetValue(locale, kCFLocaleCountryCode)));
+  const CFIndex kCStringSize = 128;
+  char temporaryCString[kCStringSize];
+  bzero(temporaryCString, kCStringSize);
```
Comment (codebytere) [inline atom/browser/api/atom_api_app.cc]: [`bzero()`](http://pubs.opengroup.org/onlinepubs/009696699/functions/bzero.html) is legacy; prefer [`memset()`](http://pubs.opengroup.org/onlinepubs/009696699/functions/memset.html)

File: docs/api/app.md [hunk 3]
```diff
@@ -580,6 +580,10 @@ To set the locale, you'll want to use a command line switch at app startup, whic
 
 **Note:** On Windows you have to call it after the `ready` events gets emitted.
 
+### `app.getRegion()` _macOS_ _Linux_ _Windows_
```
Comment (codebytere) [inline docs/api/app.md]: For functions that work across platforms, no need to specify all of `_macOS_ _Linux_ _Windows_`

**Comment by codebytere** [comment]
@zarubond could you please add spec(s) for this?

File: atom/browser/api/atom_api_app.cc [hunk 4]
```diff
@@ -1299,7 +1338,8 @@ void App::BuildPrototype(v8::Isolate* isolate,
       .SetMethod("setPath", &App::SetPath)
       .SetMethod("getPath", &App::GetPath)
       .SetMethod("setDesktopName", &App::SetDesktopName)
-      .SetMethod("getLocale", &App::GetLocale)
+      .SetMethod("getRegion", &App::GetRegion)
```
Comment (codebytere) [inline atom/browser/api/atom_api_app.cc]: `getRegion` should be removed now i believe since you changed the name

File: docs/api/app.md [hunk 5]
```diff
@@ -580,6 +580,11 @@ To set the locale, you'll want to use a command line switch at app startup, whic
 
 **Note:** On Windows you have to call it after the `ready` events gets emitted.
 
+### `app.GetLocaleCountryCode()`
+Returns `String` - User operating system´s region in ISO3166 [here](https://www.iso.org/iso-3166-country-codes.html). The value is taken from native OS APIs.
```
Comment (codebytere) [inline docs/api/app.md]: `region` => `locale country code`

File: docs/api/app.md [hunk 6]
```diff
@@ -580,6 +580,11 @@ To set the locale, you'll want to use a command line switch at app startup, whic
...
+
+**Note:** When unable to detect region, it returns empty string.
```
Comment (codebytere) [inline docs/api/app.md]: same as above ☝️

File: docs/api/app.md [hunk 7]
```diff
@@ -580,6 +580,11 @@ To set the locale, you'll want to use a command line switch at app startup, whic
 
 **Note:** On Windows you have to call it after the `ready` events gets emitted.
 
+### `app.GetLocaleCountryCode()`
```
Comment (HashimotoYT) [inline docs/api/app.md]: `app.GetLocaleCountryCode()` => `app.getLocaleCountryCode()`

**Comment by codebytere** [comment]
@zarubond spec is failing on
`AssertionError: expected 'US' to be one of [ 0, 2 ]`
you just need to correct this and we should be good to go!

**Comment by zarubond** [comment]
@codebytere Would it be possible to get output of "locale" command on Linux testing machine?

**Comment by MarshallOfSound** [comment]
@zarubond This is the output of `locale` on the linux build machine

```
locallocalbuilduser@0495b2eff748:~$ locale
LANG=
LANGUAGE=
LC_CTYPE="POSIX"
LC_NUMERIC="POSIX"
LC_TIME="POSIX"
LC_COLLATE="POSIX"
LC_MONETARY="POSIX"
LC_MESSAGES="POSIX"
LC_PAPER="POSIX"
LC_NAME="POSIX"
LC_ADDRESS="POSIX"
LC_TELEPHONE="POSIX"
LC_MEASUREMENT="POSIX"
LC_IDENTIFICATION="POSIX"
LC_ALL=
```

Doesn't look too good 😆

**Comment by zarubond** [comment]
Thanks @MarshallOfSound , so the output of the code was correct.

File: atom/browser/api/atom_api_app.cc (hunk 2 above)
Comment (miniak) [reply atom/browser/api/atom_api_app.cc]: or just `char temporaryCString[kCStringSize] = {}`

File: atom/browser/api/atom_api_app.cc [hunk 8]
```diff
@@ -882,6 +883,44 @@ std::string App::GetLocale() {
...
+#if defined(OS_WIN)
+  WCHAR locale_name[LOCALE_NAME_MAX_LENGTH] = {0};
+
+  if (GetLocaleInfoEx(LOCALE_NAME_USER_DEFAULT, LOCALE_SISO3166CTRYNAME,
+                      (LPWSTR)&locale_name,
+                      sizeof(locale_name) / sizeof(WCHAR)) ||
+      GetLocaleInfoEx(LOCALE_NAME_SYSTEM_DEFAULT, LOCALE_SISO3166CTRYNAME,
+                      (LPWSTR)&locale_name,
+                      sizeof(locale_name) / sizeof(WCHAR)))
+    base::WideToUTF8(locale_name, wcslen(locale_name), &region);
```
Comment (miniak) [inline atom/browser/api/atom_api_app.cc]: even though not mandatory, please put the body in `{}` block, it's a bit confusing now considering the condition is 6 lines of code

File: atom/browser/api/atom_api_app.cc [hunk 9]
```diff
@@ -882,6 +883,44 @@ std::string App::GetLocale() {
...
+  CFStringRef value = CFStringRef(
+      static_cast<CFTypeRef>(CFLocaleGetValue(locale, kCFLocaleCountryCode)));
+  const CFIndex kCStringSize = 128;
+  char temporaryCString[kCStringSize] = {0};
+  CFStringGetCString(value, temporaryCString, kCStringSize,
+                     kCFStringEncodingUTF8);
+  region = temporaryCString;
+#else
+  const char* locale_ptr = setlocale(LC_TIME, NULL);
+  if (locale_ptr == NULL)
```
Comment (miniak) [inline atom/browser/api/atom_api_app.cc]: `if (!locale_ptr)`

File: docs/api/app.md [hunk 10]
```diff
@@ -603,6 +603,11 @@ To set the locale, you'll want to use a command line switch at app startup, whic
 
 **Note:** On Windows you have to call it after the `ready` events gets emitted.
 
+### `app.getLocaleCountryCode()`
+Returns `String` - User operating system´s locale country code in ISO3166 [here](https://www.iso.org/iso-3166-country-codes.html). The value is taken from native OS APIs.
```
Comment (alexeykuzmin) [inline docs/api/app.md]: Lower-case `string` here please.

File: docs/api/app.md [hunk 11]
```diff
@@ -603,6 +603,11 @@ To set the locale, you'll want to use a command line switch at app startup, whic
...
+
+**Note:** When unable to detect locale country code, it returns empty string.
```
Comment (alexeykuzmin) [inline docs/api/app.md]: Is it possible to return `null`?

**Comment by alexeykuzmin** [comment]
@zarubond 
None of commits in the PR are associated with your GitHub account.
Can you please change the author of the commits or add that new email to your GitHub account?

File: docs/api/app.md (hunk 11 above)
Comment (zarubond) [reply docs/api/app.md]: There is no way how to do it from C++

File: spec/api-app-spec.js [hunk 12]
```diff
@@ -122,6 +122,12 @@ describe('app module', () => {
     })
   })
 
+  describe('app.getLocaleCountryCode()', () => {
+    it('should be empty or have length of two', () => {
+      expect(app.getLocaleCountryCode().length).to.be.oneOf([0, 2])
```
Comment (alexeykuzmin) [inline spec/api-app-spec.js]: So if the function will suddenly start returning an empty string all the time, the test won't catch it?
Is there a way to predict its return value? Does it depend on a platform?

Comment (alexeykuzmin) [reply spec/api-app-spec.js]: @zarubond 
Can you please the expected length to `0` only for the linux CI? Like this:

```js
const expectedLength = 2
if (isCI && process.platform === 'linux') {
  // Linux CI machines have no locale.
  expectedLength = 0
}
expect(app.getLocaleCountryCode()).to.be.a('string')
    .and.have.lengthOf(expectedLength)
```

Comment (alexeykuzmin) [reply spec/api-app-spec.js]: @zarubond 
Can you please set the expected length to `0` only for the linux CI? Like this:
```js
let expectedLength = 2
if (isCi && process.platform === 'linux') {
  // Linux CI machines have no locale.
  expectedLength = 0
}
expect(app.getLocaleCountryCode()).to.be.a('string')
    .and.have.lengthOf(expectedLength)
```

Comment (zarubond) [reply spec/api-app-spec.js]: Added the exception for Linux.

**Comment by alexeykuzmin** [comment]
@zarubond It should be `isCI` not `isCi`, sorry for confusion.

File: spec/api-app-spec.js [hunk 13]
```diff
@@ -122,6 +122,17 @@ describe('app module', () => {
     })
   })
 
+  describe('app.getLocaleCountryCode()', () => {
+    it('should be empty or have length of two', () => {
+      let expectedLength = 2;
```
Comment (alexeykuzmin) [inline spec/api-app-spec.js]: Electron's [JS codestyle](https://github.com/electron/electron/blob/master/docs/development/coding-style.md#javascript) doesn't allow semicolons, please remove the semicolon to fix [the lint job](https://github.visualstudio.com/electron/_build/results?buildId=20907&view=logs).

File: docs/api/app.md [hunk 14]
```diff
@@ -603,6 +603,11 @@ To set the locale, you'll want to use a command line switch at app startup, whic
 
 **Note:** On Windows you have to call it after the `ready` events gets emitted.
 
+### `app.getLocaleCountryCode()`
+Returns `string` - User operating system´s locale country code in ISO3166 [here](https://www.iso.org/iso-3166-country-codes.html). The value is taken from native OS APIs.
```
Comment (alexeykuzmin) [inline docs/api/app.md]: Please make "ISO3166" a link and remove the "here" word. The text looks weird now.
```suggestion
Returns `string` - User operating system's locale two-letter [ISO 3166](https://www.iso.org/iso-3166-country-codes.html) country code. The value is taken from native OS APIs.
```

**Pull request merged**